cookbooks/inner_core_convection/inner_core_convection.cc

```diff
@@ -0,0 +1,424 @@
+#include <aspect/global.h>
+#include <aspect/heating_model/interface.h>
+#include <aspect/simulator.h>
+#include <aspect/assembly.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/base/parsed_function.h>
+
+#include <typeinfo>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    template <int dim>
+    class InnerCore : public MaterialModel::Simple<dim>
+    {
+      public:
+        virtual void evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+                              MaterialModel::MaterialModelOutputs<dim> &out) const;
+
+        /**
+         * Constructor.
+         */
+        InnerCore ();
+
+        /**
+         * A function that is called at the beginning of each time step to
+         * indicate what the model time is for which the boundary values will
+         * next be evaluated. For the current class, the function passes to
+         * the parsed function what the current time is.
+         */
+        virtual
+        void
+        update ();
+
+        /**
+         * Declare the parameters this class takes through input files. The
+         * default implementation of this function does not describe any
+         * parameters. Consequently, derived classes do not have to overload
+         * this function if they do not take any runtime parameters.
+         */
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Read the parameters this class declares from the parameter file.
+         * The default implementation of this function does not read any
+         * parameters. Consequently, derived classes do not have to overload
+         * this function if they do not take any runtime parameters.
+         */
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+        /**
+         * A function object representing resistance to phase change at the
+         * inner core boundary.
+         */
+        Functions::ParsedFunction<dim> resistance_to_phase_change;
+    };
+
+  }
+}
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+
+    template <int dim>
+    void
+    InnerCore<dim>::
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      // First, we use the material descriptions of the 'simple' material model to fill all of the material
+      // model outputs. Below, we will then overwrite selected properties (the specific heat) to make the
+      // product of density and specific heat a constant.
+      Simple<dim>::evaluate(in, out);
+
+      // We want the right-hand side of the momentum equation to be (- Ra T gravity) and
+      // density * cp to be 1
+      for (unsigned int q=0; q < in.position.size(); ++q)
+        {
+          out.densities[q] = - out.thermal_expansion_coefficients[q] * (in.temperature[q]);
```

Comment:
The negative density works here because it is only on the right-hand side of the momentum equation (it's the Boussinesq approximation, so the density is constant in the temperature equation).